presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -46,20 +49,32 @@
     private final List<PlanOptimizer> legacyRules;
     private final RuleIndex ruleIndex;
     private final StatsRecorder stats;
+    private final RuleApplicationListener ruleApplicationListener;
 
     public IterativeOptimizer(StatsRecorder stats, Set<Rule<?>> rules)
     {
         this(stats, ImmutableList.of(), rules);
     }
 
+    public IterativeOptimizer(StatsRecorder stats, RuleApplicationListener ruleApplicationListener, Set<Rule<?>> rules)
+    {
+        this(stats, ruleApplicationListener, ImmutableList.of(), rules);
+    }
+
     public IterativeOptimizer(StatsRecorder stats, List<PlanOptimizer> legacyRules, Set<Rule<?>> newRules)
     {
+        this(stats, noListener(), legacyRules, newRules);
+    }
+
+    public IterativeOptimizer(StatsRecorder stats, RuleApplicationListener ruleApplicationListener, List<PlanOptimizer> legacyRules, Set<Rule<?>> newRules)
+    {
         this.legacyRules = ImmutableList.copyOf(legacyRules);
         this.ruleIndex = RuleIndex.builder()
                 .register(newRules)
                 .build();
 
         this.stats = stats;
+        this.ruleApplicationListener = ruleApplicationListener;
```

Comment:
`requireNonNull`. However, I don't like the fact Iterative's internals are so exposed to the outer world, this will make more coupled and harder to change

Reply:
None of the structures that are part of callback are specific to `IterativeOptimizer`. Do you mean order in which `IterativeOptimizer` executes the rules? I think we should only rely on the fact that iterative optimizer is exhaustive and write tests based on that. We shouldn't assume any particular order if it isn't enforced by rule and trait design. Therefore we can only tests things like:
* rule executes at most once/doesn't loop forever
* execution of rule X should be before execution of rule Y only if there is strong dependency between the rules (e.g: rule X produces trait or specific plan node that rule Y requires).

We cannot test things like:
* rule X executes before Y if there is not strong dependency between them (e.g: both match pattern)